Spatial plotting post

crwilliams11 opened the pull request:
This post shows a number of plots that can be generated from Atomx SIP-exported flat files as well as the RNA QC custom module. The goal is for it to be used as inspiration for customers to start plotting their spatial data. I imagine that we can add additional plotting examples to it in the future if there are additional key plots, color palettes, etc. that we'd like to show.

crwilliams11 committed:
initial draft of plotting post

crwilliams11 committed:
add author, update legend for alpha, add one FOV metric

patrickjdanaher commented:
Really nice and comprehensive. 

General comments:
- one more plot to consider: a plot faceted by cell type. E.g. 20 panels, each panel showing one cell type in a dark color and the others in light grey. 
- looks like you're using cell_meta$CenterX_global_px and converting to mm. Why not just use slide_mm_x?
- consider some comments in the ggplot code. e.g.,
```
  annotation_custom(                                           # add scale bar
    grob = rectGrob(gp = gpar(fill = "black")),
    xmin = x_start, xmax = x_end,
    ymin = y_pos, ymax = y_pos + scale_length_px*0.05
  ) +
```
- there are so many lines of code taken up by the scale bars. could you condense them somehow? e.g., create a function that taked in the xrange and yrange and outputs 1. the polygon object and 2. the text annotation object? Then call ggplot() + ... + annotation_custom(rectangle_object)...? (You know, writing this out, it seems awkward; either come up with something more elegant or ignore this idea.)
- ggplot's default colors are really unhelpful - they just blend indistinguishably into each other. you should show use of a custom color palette. I like to ask chatGPT for a vector of 20 or 30 highly distinct hexcodes. 
- ggplot of polygons: can you try making the borders thinner? OR, it might look nice to use some alpha (say 0.6?) in the cell interiors then color the borders the same, but without transparency, 
- python polygon plots would probably benefit from borders, though not essential
- border and fill different colors: do we want to recommend this? I find them hard to parse. Instead, can you show the following? Mute all cell types but one, and color it dynamically. E.g. color only tumor by total counts, and leave everything else "grey80" or something similarly muted. (this plot might go more naturally later in the article). 
- highlighting one cell type: in R, can you use a similar light grey + rich color  duo as you did in py?
- "mute background cells": I've had really good luck coloring the selected cell type black, which lets you mute the background cells a little less. Worth trying / considering. 
- plotting flagged cells: is it possible to plot a spatially smoothed version of this, e.g. color each cell by %flagged in its 50 nearest neighbors? as is, I think it's a little misleading, since with the flagged cells on top it looks like some areas are completely lost.

crwilliams11 committed:
changes per PR comments

crwilliams11 commented:
Today's commit addresses the above PR comments as follows:

1. _one more plot to consider: a plot faceted by cell type. E.g. 20 panels, each panel showing one cell type in a dark color and the others in light grey._

Added as section 3.7

2. _looks like you're using cell_meta$CenterX_global_px and converting to mm. Why not just use slide_mm_x?_

Default flat files do not contain the mm columns and this post is designed to work from the flat files. I added this note to the scale bar section: `Additionally, all the code here is designed to work from positions given in pixels, as they are in default flat files. If you have positions in mm you'll want to adjust accordingly, primarily in the scale bar function.`

3. _consider some comments in the ggplot code._

I've added a few throughout; open to adding more if you think there are specific sections where they'd be helpful.

4. _there are so many lines of code taken up by the scale bars. could you condense them somehow? e.g., create a function that taked in the xrange and yrange and outputs 1. the polygon object and 2. the text annotation object? Then call ggplot() + ... + annotation_custom(rectangle_object)...? (You know, writing this out, it seems awkward; either come up with something more elegant or ignore this idea.)_

Good idea. I've added a scale bar function to load at the top and then call this in each subsequent spatial plot, which cleans up the code a lot.

5. _ggplot's default colors are really unhelpful - they just blend indistinguishably into each other. you should show use of a custom color palette. I like to ask chatGPT for a vector of 20 or 30 highly distinct hexcodes._

I agree that the ggplot defaults are often unhelpful. I now instead show multiple existing palette options, with most plots demonstrating a separate palette. I added this note: `Throughout the post, we showcase a variety of color palettes. When only a few colors are needed, we aim for consistency between R and Python plots. In cases requiring many colors, we highlight diverse palette options, which may differ across the two languages.`

6. _ggplot of polygons: can you try making the borders thinner? OR, it might look nice to use some alpha (say 0.6?) in the cell interiors then color the borders the same, but without transparency,_

Done. I went with making the borders thinner; adding transparency when you have 20+ colors makes the colors harder to distinguish.

7. _python polygon plots would probably benefit from borders, though not essential_

Added.

8. _border and fill different colors: do we want to recommend this? I find them hard to parse. Instead, can you show the following? Mute all cell types but one, and color it dynamically. E.g. color only tumor by total counts, and leave everything else "grey80" or something similarly muted. (this plot might go more naturally later in the article)._

I also find them a bit hard to parse. I've added your suggested plot as section 3.6 instead.

9. _highlighting one cell type: in R, can you use a similar light grey + rich color duo as you did in py?_

I've made both languages use the same color pair now.

10. _"mute background cells": I've had really good luck coloring the selected cell type black, which lets you mute the background cells a little less. Worth trying / considering._

I added this as a suggestion in the text, `As an alternative, we can highlight one cell type by muting the color for all others. However, keep in mind that when many colors are present, muting can reduce contrast and make distinctions harder to perceive. To improve interpretability, consider grouping cells into 5–10 major types or manually assigning similar hues to related cell types. As another option, color the cells of interest black, which may allow you to mute other colors less.`

11. _plotting flagged cells: is it possible to plot a spatially smoothed version of this, e.g. color each cell by %flagged in its 50 nearest neighbors? as is, I think it's a little misleading, since with the flagged cells on top it looks like some areas are completely lost._

I prefer to show all flagged cells to clearly show where they're occurring. However, I added a note about overplotting and increased the opacity so that you still get some of that regional density effect visually. `Keep an eye out for overplotting here, which can lead to the incorrect conclusion that all cells in a region have been lost. We reduce the opacity to counter this impression.`

crwilliams11 committed:
change thumbnail picture

crwilliams11 committed:
Merge branch 'Main' into useful-plotting
